### PR TITLE
CORE: Validate list values for settings

### LINKED ISSUE #33135: String settings leniently accept lists
For example, the following `elasticsearch.yml` ...

```
node.name:
- node-a
- node-b
```

... starts up a node called `[node-a, node-b]`:

```
[2018-08-24T17:08:43,807][INFO ][o.e.n.Node               ] [[node-a, node-b]] initializing ...
```

In fact this is what's happening in #30946. The following API call...

```
PUT _cluster/settings
{
  "transient" : {
    "cluster.routing.allocation.exclude._name" : [ "elasticsearch-data-0" ]
  }
}
```

... sets `cluster.routing.allocation.exclude._name` to the string `[elasticsearch-data-0]`, which then does not match the node called `elasticsearch-data-0` so the allocation is not filtered as expected.

### PR DESCRIPTION
**Pull request by amitds1997**
Settings value could be a list. This should only
happen if the underlying setting type is a list setting type.

* Validation has been added that
	* when we get a setting value which is list, setting
is also a list setting
	* when we get a value for a list setting, underlying
 value is also a list
* Closes #33135

### CODE REVIEW

**Comment by pgomulka** [comment]
ok to test

**Comment by pgomulka** [comment]
ok to test

**Comment by pgomulka** [comment]
@amitds1997 this looks good, but can you fix the test? 
You might need to change the message and get rid of the check in Environment 
```
 String rawDataPath = PATH_DATA_SETTING.get(settings);
            if (rawDataPath.startsWith("[")) {
```

can you also add a testcase for SettingsTests with a happy path where a `Setting.listSetting` successfully parses ?

File: server/src/test/java/org/elasticsearch/common/settings/SettingsTests.java [hunk 1]
```diff
@@ -492,6 +492,13 @@ public void testSecureSettingIllegalName() {
...
     }
 
+    public void testValidateStringSetting() {
+      Settings settings = Settings.builder().putList("foo.bar", Arrays.asList("bla-a", "bla-b")).build();
+      Setting<String> stringSetting = Setting.simpleString("foo.bar");
+      IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () -> stringSetting.get(settings));
+      assertEquals("Found list type value for setting [foo.bar] but did not expect a list for it.", e.getMessage());
+    }
+
     public void testGetAsArrayFailsOnDuplicates() {
```
Comment (pgomulka) [inline server/src/test/java/org/elasticsearch/common/settings/SettingsTests.java:502]: can we also have a happy path testcase where `Setting.listSetting` parses successfully ?

**Review by pgomulka — CHANGES_REQUESTED** [review]
looks good, thanks!
can you add the testcase for happy path and fix the failing test?
You might need to change the exception message and remove the if condition that throws Environment
```
 String rawDataPath = PATH_DATA_SETTING.get(settings);
            if (rawDataPath.startsWith("[")) {
```

**Comment by rjernst** [comment]
@amitds1997 Are you still interested in getting this PR in? If so can you please address the comments Przemek made above?

**Comment by amitds1997** [comment]
Sorry, I won't be able to get this through. Anyone else can take this up. Thanks!

**Pull request closed without merging**